tests/runtime/test_command_buffer.c

```diff
@@ -0,0 +1,186 @@
+
+  const char *code = STR (kernel void vector_addition (
+      global int *tile1, global int *tile2, global int *res) {
+    size_t index = get_global_id (0);
+    res[index] = tile1[index] + tile2[index];
+  });
+  const size_t length = strlen (code);
+
+  cl_program program
+      = clCreateProgramWithSource (context, 1, &code, &length, &error);
+  CHECK_CL_ERROR (error);
+  CHECK_CL_ERROR (clBuildProgram (program, 1, &device, NULL, NULL, NULL));
+  cl_kernel kernel = clCreateKernel (program, "vector_addition", &error);
+  CHECK_CL_ERROR (error);
+
+  size_t frame_count = 60;
+  size_t frame_elements = 1024;
+  size_t frame_size = frame_elements * sizeof (cl_int);
+
+  size_t tile_count = 16;
+  size_t tile_elements = frame_elements / tile_count;
+  size_t tile_size = tile_elements * sizeof (cl_int);
+
+  cl_mem buffer_tile1
+      = clCreateBuffer (context, CL_MEM_READ_ONLY, tile_size, NULL, &error);
+  CHECK_CL_ERROR (error);
+  cl_mem buffer_tile2
+      = clCreateBuffer (context, CL_MEM_READ_ONLY, tile_size, NULL, &error);
+  CHECK_CL_ERROR (error);
+  cl_mem buffer_res
+      = clCreateBuffer (context, CL_MEM_WRITE_ONLY, tile_size, NULL, &error);
+  CHECK_CL_ERROR (error);
+
+  CHECK_CL_ERROR (
+      clSetKernelArg (kernel, 0, sizeof (buffer_tile1), &buffer_tile1));
+  CHECK_CL_ERROR (
+      clSetKernelArg (kernel, 1, sizeof (buffer_tile2), &buffer_tile2));
+  CHECK_CL_ERROR (
+      clSetKernelArg (kernel, 2, sizeof (buffer_res), &buffer_res));
+
+  cl_command_queue command_queue = clCreateCommandQueue (
+      context, device, CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE, &error);
+  CHECK_CL_ERROR (error);
+
+  cl_command_buffer_khr command_buffer
+      = ext.clCreateCommandBufferKHR (1, &command_queue, NULL, &error);
+  CHECK_CL_ERROR (error);
+
+  cl_mem buffer_src1
+      = clCreateBuffer (context, CL_MEM_READ_ONLY, frame_size, NULL, &error);
+  CHECK_CL_ERROR (error);
+  cl_mem buffer_src2
+      = clCreateBuffer (context, CL_MEM_READ_ONLY, frame_size, NULL, &error);
+  CHECK_CL_ERROR (error);
+  cl_mem buffer_dst
+      = clCreateBuffer (context, CL_MEM_READ_WRITE, frame_size, NULL, &error);
+  CHECK_CL_ERROR (error);
+
+  cl_sync_point_khr tile_sync_point = 0;
+  for (size_t tile_index = 0; tile_index < tile_count; tile_index++)
+    {
+      cl_sync_point_khr copy_sync_points[2];
+      CHECK_CL_ERROR (ext.clCommandCopyBufferKHR (
+          command_buffer, NULL, buffer_src1, buffer_tile1,
+          tile_index * tile_size, 0, tile_size, tile_sync_point ? 1 : 0,
+          tile_sync_point ? &tile_sync_point : NULL, &copy_sync_points[0],
+          NULL));
+      CHECK_CL_ERROR (ext.clCommandCopyBufferKHR (
+          command_buffer, NULL, buffer_src2, buffer_tile2,
+          tile_index * tile_size, 0, tile_size, tile_sync_point ? 1 : 0,
+          tile_sync_point ? &tile_sync_point : NULL, &copy_sync_points[1],
+          NULL));
+
+      cl_sync_point_khr nd_sync_point;
+      CHECK_CL_ERROR (ext.clCommandNDRangeKernelKHR (
```

Comment:
This tests only the NDRange cmd. Can you add (smoke/unit) tests for the rest too?